cpp/src/gandiva/parse_url_holder.h

```diff
@@ -0,0 +1,145 @@
+    ~ParseUrlHolder() override = default;
+
+    // Invoked by function_holder_registry.h
+    static Status Make(const FunctionNode& node, std::shared_ptr<ParseUrlHolder>* holder);
+
+    static Status Make(const std::string& extract_part, std::shared_ptr<ParseUrlHolder>* holder);
+
+    const char * operator()(ExecutionContext *ctx, const char * url, int32_t url_len, int32_t *out_length) {
+        std::string url_string(url, url_len);
+        arrow::internal::Uri uri;
+        std::string out;
+
+        auto status = uri.Parse(url_string);
+        if (!status.ok()) {
+            return nullptr;
+        }
+
+        if (extract_part_ == "HOST") {
+            out = uri.host();
+        } else if (extract_part_ == "PATH") {
+            out = uri.path();
+        } else if (extract_part_ == "QUERY") {
+            out = uri.query_string();
+        } else if (extract_part_ == "PROTOCOL") {
+            out = uri.scheme();
+        } else if (extract_part_ == "FILE") {
+            if (uri.has_query()) {
+                out = uri.path() + "?" + uri.query_string();
+            } else {
+                out = uri.path();
+            }
+        } else if (extract_part_ == "AUTHORITY") {
+            if (uri.has_port()) {
+                out = uri.host() + ":" + uri.port_text();
+            } else {
+                out = uri.host();
+            }
+        } else if (extract_part_ == "USERINFO") {
+            out = uri.user_info();
+        } else {
+            return nullptr;
+        }
+
+        *out_length = static_cast<int32_t>(out.length());
+        if (*out_length == 0) {
+            return nullptr;
+        }
+
+        char *result_buffer = reinterpret_cast<char *>(ctx->arena()->Allocate(*out_length));
+        if (result_buffer == NULLPTR) {
+            ctx->set_error_msg("Could not allocate memory for result! Wrong result may be returned!");
+            *out_length = 0;
+            return nullptr;
+        }
+        memcpy(result_buffer, out.data(), *out_length);
+
+        return result_buffer;
+    }
+
+    const char * operator()(ExecutionContext *ctx,
+                            const char * url, int32_t url_len,
+                            const char * pattern, int32_t pattern_len, int32_t *out_length) {
+        std::string url_string(url, url_len);
+        std::string pattern_string(pattern, pattern_len);
+        arrow::internal::Uri uri;
+        std::string out;
+
+        auto status = uri.Parse(url_string);
+        if (!status.ok()) {
+            return nullptr;
+        }
+
+        if (extract_part_ != "QUERY" || !uri.has_query()) {
+            return nullptr;
+        } else {
+            std::unordered_map<std::string, std::string> queries;
+            const auto items = std::move(uri.query_items()).ValueUnsafe();
+            for (const auto& query : items) {
+                queries.emplace(query.first, query.second);
+            }
+
+            auto out_query = queries.find(pattern_string);
```

Comment:
It looks like pattern is viewed as a regex in the implementation of vanilla spark. Please check this part further.
If needed, you can use `re2` lib to find the matched part. Please refer to `extract_holder.h` to see how to call the APIs in `re2`.

Reply:
It seems that spark uses regexp to obtain query data, and there is no regular-related case in the test case. I'm not sure if it's necessary to support regex here.

Reply:
OK. We can keep the current impl to just support plain string in this PR. Please leave a comment to clarify this.

Reply:
Sure.